packages/nimble-components/src/table-column/enum-base/index.ts

```diff
@@ -0,0 +1,132 @@
+    TableStringField,
+    TableBooleanField,
+    TableNumberField
+} from '../../table/types';
+import { TableColumn } from '../base';
+import { Mapping } from '../../mapping/base';
+import type { MappingKeyType } from './types';
+import type { MappingConfig } from './models/mapping-config';
+import type { MappingKey } from '../../mapping/base/types';
+import { resolveKeyWithType } from './models/mapping-key-resolver';
+import type { TableColumnEnumBaseValidator } from './models/table-column-enum-base-validator';
+
+export type TableColumnEnumCellRecord =
+    | TableStringField<'value'>
+    | TableBooleanField<'value'>
+    | TableNumberField<'value'>;
+export type MappingConfigs = Map<MappingKey, MappingConfig>;
+export interface TableColumnEnumColumnConfig {
+    mappingConfigs: MappingConfigs;
+}
+
+/**
+ * Base class for table columns that map values to content (e.g. nimble-table-column-enum-text and nimble-table-column-icon)
+ */
+export abstract class TableColumnEnumBase<
+    TColumnConfig extends TableColumnEnumColumnConfig,
+    TEnumValidator extends TableColumnEnumBaseValidator<[]>
+>
+    extends TableColumn<TColumnConfig>
+    implements Subscriber {
+    // To ensure the validator is available when other properties get initialized
+    // (which can trigger validation), declare the validator first.
+    /** @internal */
+    public validator = this.createValidator();
+
+    /** @internal */
+    public mappingNotifiers: Notifier[] = [];
+
+    /** @internal */
+    @observable
+    public mappings: Mapping[] = [];
+
+    @attr({ attribute: 'field-name' })
+    public fieldName?: string;
+
+    @attr({ attribute: 'key-type' })
+    public keyType: MappingKeyType = 'string';
+
+    /**
+     * @internal
+     *
+     * Triggers a request to update the columnConfig when any observable property on
+     * a mapping is updated.
+     */
+    public handleChange(source: unknown, args: unknown): void {
+        if (source instanceof Mapping && typeof args === 'string') {
+            this.updateColumnConfig();
+        }
+    }
+
+    public abstract createValidator(): TEnumValidator;
+
+    /**
+     * Called when any Mapping related state has changed.
+     * Implementations should run validation before updating the column config.
+     */
+    protected abstract updateColumnConfig(): void;
+
+    /**
+     * Implementations should throw an error if an invalid Mapping is passed.
+     */
+    protected abstract createMappingConfig(mapping: Mapping): MappingConfig;
+
+    /**
+     * Validation state should be updated before calling this function.
+     */
+    protected createColumnConfig(): TableColumnEnumColumnConfig | undefined {
+        if (!this.validator.isValid) {
+            return undefined;
+        }
+        const mappingConfigs = new Map<MappingKey, MappingConfig>();
+        this.mappings.forEach(mapping => {
+            const key = resolveKeyWithType(mapping.key, this.keyType);
+            if (key !== undefined) {
```

Comment:
Given all the other updates, would it make sense now to remove this `key !== undefined` check? Or, do you think it is still beneficial to keep it?

Reply:
I still think it's okay to have, especially because I would otherwise need to add a non-null assertion anyway.

Reply:
I think this can at least behave consistently with createMappingConfig. The idea being this is an impossible state since we have guarded the running of this function with validation so we should throw instead.